### PR TITLE
[ROCm] Add support for Punica kernels on AMD GPUs

### PR DESCRIPTION
**Pull request by kliuae**
This PR adds ROCm support for punica kernels to enable multi-LoRA on AMD GPUs.
Some Punica files are slightly refactored so that the correct c++/hipcc compilers can be invoked when building under ROCm.
A custom bgmv shrink kernel is added to account for the difference in warp size between AMD's GPUs and Nvidia's.
The port has been tested on MI210, and the unit tests applying LoRA are passing.

### CODE REVIEW

**Comment by WoosukKwon** [comment]
@hongxiayang @lcskrishna Could you help review this PR?

**Comment by WoosukKwon** [comment]
@hongxiayang @dllehr-amd Could you review this PR? This is an important PR that enables the AMD GPUs to support multi-LoRA serving, which is a key feature in vLLM liked by many users.

**Comment by simon-mo** [comment]
This script can help verify this works end to end https://github.com/vllm-project/vllm/blob/main/examples/multilora_inference.py

**Comment by hongxiayang** [comment]
> This script can help verify this works end to end https://github.com/vllm-project/vllm/blob/main/examples/multilora_inference.py

thanks @simon-mo . this scripts needs `from vllm.lora.request import LoRARequest` which is not available in this pull request.

**Comment by hongxiayang** [comment]
will check. Thanks for this effort.

**Review by hongxiayang — APPROVED** [review]
I did some verification and ran the e2e verify script.

At first punica C extension was not build by default and import failed. And then I added the env VLLM_INSTALL_PUNICA_KERNELS and then rebuild the docker image. It was ok afterwards. Maybe worth a mention in the documentation somewhere if not already.

File: csrc/punica/type_convert.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+#ifndef USE_ROCM
+
+#include <cuda_bf16.h>
+#include <cuda_fp16.h>
+
+#else
+
+#include <hip/hip_bf16.h>
+#include <hip/hip_fp16.h>
```
[the remark this answers is not in the document]
Comment (kliuae) [reply csrc/punica/type_convert.h:12]: Ideally we should not be needing this, but we may still need this for the time being as in quite a few environments I've tested with, including the rocm/pytorch images Dockerfile.rocm builds from, the hipify tool seems to have failed to convert the bf16 header.

File: csrc/punica/punica_pybind.cpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#include <torch/extension.h>
+
+#include "punica_ops.h"
+
+//====== pybind ======
+
+#define DEFINE_pybind(name) m.def(#name, &name, #name);
```
[the remark this answers is not in the document]
Comment (kliuae) [reply csrc/punica/punica_pybind.cpp:7]: This is not used but it is shipped with the implementation for cuda, so I kept it when refactoring for consistency.

[the remark this answers is not in the document]
Comment (hongxiayang) [reply csrc/punica/punica_pybind.cpp:7]: @kliuae  please ping @simon-mo  or @WoosukKwon to further review/approve this PR so that it can be merged.

**Comment by jamestwhedbee** [comment]
I was going to try this out soon, is this in a good spot or is it still being worked?

**Comment by kliuae** [comment]
> I was going to try this out soon, is this in a good spot or is it still being worked?

It's in a good state for testing, though occasionally I'll be merging in the upstream to fix conflicts before it gets merged.

**Comment by WoosukKwon** [comment]
@kliuae Sorry for the late review. The PR looks good. Could you please resolved the merge conflict in `CMakeLists.txt` so that I can merge it? Thanks!

**Review by WoosukKwon — APPROVED** [review]
@kliuae Sorry for the late review. The PR looks good. Could you please resolved the merge conflict in CMakeLists.txt so that I can merge it? Thanks!

**Comment by Alexei-V-Ivanov-AMD** [comment]
> @kliuae Sorry for the late review. The PR looks good. Could you please resolved the merge conflict in CMakeLists.txt so that I can merge it? Thanks!

@kliuae Please resolve the latest merge conflict. Your PR is instrumental for our ongoing effort. Thank you very much!

**Comment by kliuae** [comment]
@WoosukKwon Merge conflicts are resolved

**Pull request merged**